src/execution_plan/execution_plan.c

```diff
@@ -799,11 +799,9 @@ ExecutionPlan *NewExecutionPlan(RedisModuleCtx *ctx, GraphContext *gc, ResultSet
 	// Optimize the operations in the ExecutionPlan.
 	optimizePlan(gc, plan);
 
-	// Prepare the result set if this query returns records.
-	if(result_set) {
-		result_set->exps = segment->projections;
-		ResultSet_ReplyWithPreamble(result_set, segment->query_graph);
-	}
+	// Prepare the result set and emit a header if this query returns records.
+	if(result_set) ResultSet_ReplyWithPreamble(result_set, segment->query_graph, segment->projections);
```

Comment:
Just an idea, how about we introduce a ResultSet_AddColumn function
which will be called for every projected expression,
if the result-set doesn't have any columns it won't emit a header and won't emit records.